### DISCUSSION TITLE
Service Endpoint deployment clarification

### DISCUSSION THREAD

**Discussion by kevball2** [brwilkinson/AzureDeploymentFramework#13] [Q&A] [answered]
Wanted to see if you could clarify what was happening with the service endpoint property as part of a VNET deployment 

I am not following how the service endpoint information from the parameters file will update the ServiceEndpoints variable with the right information

```bicep
serviceEndpoints: contains(sn, 'serviceEndpoints') ? serviceEndpoints[sn.serviceEndpoints] : serviceEndpoints.default
```
the sn,ServiceEndpoints value being passed is a string, how is that replacing the default value? 

```bicep

var serviceEndpoints = {
  default: []
  'Microsoft.Storage': [
    {
      service: 'Microsoft.Storage'
      locations: [
        resourceGroup().location
      ]
    }
  ]
}
```

**Comment by brwilkinson** [comment, marked as answer]
Hi @kevball2 

sorry for delay, I didn't see this one.

This is simply a lookup, however the name of the key for the lookup is the resource type.

https://github.com/brwilkinson/AzureDeploymentFramework/blob/main/ADF/tenants/AOA/ACU1.T5.parameters.json#L359

```json
          {
            "name": "snBE02",
            "prefix": "64/28",
            "NSG": 1,
            "Route": 0, //1
            "FlowLogEnabled": 1,
            "FlowAnalyticsEnabled": 1,
            "delegations": "Microsoft.ContainerInstance/containerGroups"
          }
```

```bicep
var delegations = {
  default: []
  'Microsoft.Web/serverfarms': [
    {
      name: 'delegation'
      properties: {
        serviceName: 'Microsoft.Web/serverfarms'
      }
    }
  ]
  'Microsoft.ContainerInstance/containerGroups': [
    {
      name: 'delegation'
      properties: {
        serviceName: 'Microsoft.ContainerInstance/containerGroups'
      }
    }
  ]
  'Microsoft.Network/dnsResolvers': [
    {
      name: 'delegation'
      properties: {
        serviceName: 'Microsoft.Network/dnsResolvers'
      }
    }
  ]
}

 delegations: contains(sn, 'delegations') ? delegations[sn.delegations] : delegations.default
```

It's easier to show with the delegations above. Most subnets can only handle the 1 delegation, however if they support more you can potentially have multiple resources types that get assigned.

Basically if you don't specify any it just used the blank array from default.

**Reply by brwilkinson** [reply]
you could just use any name for the key, however I mapped it 1 to 1 with the name.

I think i considered using the `serverfarm` resource type, however I found that was not unique enough, so just went wit the full providertype.

**Reply by brwilkinson** [reply]
I would add, it would be best to use private link than service endpoint at this point.

**Reply by kevball2** [reply]
Thanks for the follow up @brwilkinson , the default part was tripping me up. I wasn't looking at the delegations var as an object. I had not seen that combination of values before. 
I agree private link is the way to go. Have you looked at network manager yet for deploying connectivity? I tried using it but didn't feel like it made deploying with Infra as Code any easier
